File: client/src/game/engine/weaponEngagement.ts
import { randomUUID } from "@/utils/generateUUID";
import { fromLonLat, get as getProjection } from "ol/proj";
import {
  DEFAULT_OL_PROJECTION_CODE,
  NAUTICAL_MILES_TO_METERS,
} from "@/utils/constants";
import Aircraft from "@/game/units/Aircraft";
import Facility from "@/game/units/Facility";
import { Circle } from "ol/geom";
import Scenario from "@/game/Scenario";
import Weapon from "@/game/units/Weapon";
import {
  getBearingBetweenTwoPoints,
  getDistanceBetweenTwoPoints,
  getNextCoordinates,
  getTerminalCoordinatesFromDistanceAndBearing,
  randomFloat,
  randomInt,
} from "@/utils/mapFunctions";
import Airbase from "@/game/units/Airbase";
import Ship from "@/game/units/Ship";
import SimulationLogs, { SimulationLogType } from "@/game/log/SimulationLogs";

export type Target = Aircraft | Facility | Weapon | Airbase | Ship;

const WEAPON_ENDGAME_DISTANCE_KM = 1;

// 击毁敌方单位的基础分值（攻方 Side.totalScore 累加）。
// 数值参考 wargame 习惯：高价值固定目标 > 舰艇 > 防空设施 > 飞机 > 武器拦截。
export const UNIT_SCORE_TABLE = {
  aircraft: 10,
  weapon: 1,
  facility: 30,
  ship: 50,
  airbase: 100,
} as const;

// 关键单位（isObjective）被毁的额外加分；同时会让 Game.checkGameEnded
// 立刻把对方判为胜方。
export const OBJECTIVE_BONUS_SCORE = 200;

// 在 weaponEndgame 命中后调用：给攻方加分，并在 isObjective 命中时记录事件
// 到 Scenario 上，由 Game 在 step 末尾统一读取并触发胜负判定。
export function onUnitDestroyed(
  currentScenario: Scenario,
  attackerSideId: string,
  target: Target
) {
  const attackerSide = currentScenario.getSide(attackerSideId);
  if (!attackerSide) return;

  let baseScore = 0;
  let unitType: "aircraft" | "weapon" | "facility" | "ship" | "airbase" =
    "weapon";
  if (target instanceof Aircraft) {
    baseScore = UNIT_SCORE_TABLE.aircraft;
    unitType = "aircraft";
  } else if (target instanceof Weapon) {
    baseScore = UNIT_SCORE_TABLE.weapon;
    unitType = "weapon";
  } else if (target instanceof Facility) {
    baseScore = UNIT_SCORE_TABLE.facility;
    unitType = "facility";
  } else if (target instanceof Ship) {
    baseScore = UNIT_SCORE_TABLE.ship;
    unitType = "ship";
  } else if (target instanceof Airbase) {
    baseScore = UNIT_SCORE_TABLE.airbase;
    unitType = "airbase";
  }

  const isObjective =
    !(target instanceof Weapon) && (target as { isObjective?: boolean }).isObjective === true;
  const totalScore = baseScore + (isObjective ? OBJECTIVE_BONUS_SCORE : 0);

  attackerSide.totalScore = (attackerSide.totalScore ?? 0) + totalScore;

  if (isObjective) {
    currentScenario.lastObjectiveDestroyed = {
      attackerSideId,
      victimSideId: (target as { sideId: string }).sideId,
      unitId: target.id,
      unitName: target.name,
      unitType,
      destroyedAt: currentScenario.currentTime,
    };
  }
}

export function isThreatDetected(
  threat: Aircraft | Weapon,
  detector: Facility | Ship | Aircraft
): boolean {
  const projection = getProjection(DEFAULT_OL_PROJECTION_CODE);
  const detectorRangeGeometry = new Circle(
    fromLonLat([detector.longitude, detector.latitude], projection!),
    detector.getDetectionRange() * NAUTICAL_MILES_TO_METERS
  );
  return detectorRangeGeometry.intersectsCoordinate(
    fromLonLat([threat.longitude, threat.latitude], projection!)
  );
}

export function weaponCanEngageTarget(target: Target, weapon: Weapon) {
  const weaponEngagementRangeNm = weapon.getEngagementRange();
  const distanceToTargetKm = getDistanceBetweenTwoPoints(
    weapon.latitude,
    weapon.longitude,
    target.latitude,
    target.longitude
  );
  const distanceToTargetNm =
    (distanceToTargetKm * 1000) / NAUTICAL_MILES_TO_METERS;
  return distanceToTargetNm <= weaponEngagementRangeNm;
}

export function checkTargetTrackedByCount(
  currentScenario: Scenario,
  target: Target
) {
  let count = 0;
  currentScenario.weapons.forEach((weapon) => {
    if (weapon.targetId === target.id) count += 1;
  });
  return count;
}

export function weaponEndgame(
  currentScenario: Scenario,
  weapon: Weapon,
  target: Target,
  simulationLogs: SimulationLogs
): boolean {
  currentScenario.weapons = currentScenario.weapons.filter(
    (currentScenarioWeapon) => currentScenarioWeapon.id !== weapon.id
  );
  if (randomFloat() <= weapon.lethality) {
    if (target instanceof Aircraft) {
      currentScenario.aircraft = currentScenario.aircraft.filter(
        (currentScenarioAircraft) => currentScenarioAircraft.id !== target.id
      );
    } else if (target instanceof Facility) {
      currentScenario.facilities = currentScenario.facilities.filter(
        (currentScenarioFacility) => currentScenarioFacility.id !== target.id
      );
    } else if (target instanceof Weapon) {
      currentScenario.weapons = currentScenario.weapons.filter(
        (currentScenarioWeapon) => currentScenarioWeapon.id !== target.id
      );
    } else if (target instanceof Airbase) {
      currentScenario.airbases = currentScenario.airbases.filter(
        (currentScenarioAirbase) => currentScenarioAirbase.id !== target.id
      );
    } else if (target instanceof Ship) {
      currentScenario.ships = currentScenario.ships.filter(
        (currentScenarioShip) => currentScenarioShip.id !== target.id
      );
    }
    onUnitDestroyed(currentScenario, weapon.sideId, target);
    simulationLogs.addLog(
      weapon.sideId,
      `${weapon.name} has hit and destroyed ${target.name}`,
      currentScenario.currentTime,
      SimulationLogType.WEAPON_HIT
    );
    return true;
  }
  simulationLogs.addLog(
    weapon.sideId,
    `${weapon.name} has missed ${target.name}`,
    currentScenario.currentTime,
    SimulationLogType.WEAPON_MISSED
  );
  return false;
}

export function launchWeapon(
  currentScenario: Scenario,
  origin: Aircraft | Facility | Ship,
  target: Target,
  launchedWeapon: Weapon,
  launchedWeaponQuantity: number,
  simulationLogs: SimulationLogs
) {
  if (
    origin.weapons.length === 0 ||
    launchedWeapon.currentQuantity < launchedWeaponQuantity
  )
    return;

  for (let i = 0; i < launchedWeaponQuantity; i++) {
    const nextWeaponCoordinates = getNextCoordinates(
      origin.latitude,
      origin.longitude,
      target.latitude,
      target.longitude,
      launchedWeapon.speed
    );
    const nextWeaponLatitude = nextWeaponCoordinates[0];
    const nextWeaponLongitude = nextWeaponCoordinates[1];
    const newWeapon = new Weapon({
      id: randomUUID(),
      name: `${launchedWeapon.name} #${randomInt(1000)}`,
      sideId: origin.sideId,
      className: launchedWeapon.className,
      latitude: nextWeaponLatitude,
      longitude: nextWeaponLongitude,
      altitude: launchedWeapon.altitude,
      heading: getBearingBetweenTwoPoints(
        nextWeaponLatitude,
        nextWeaponLongitude,
        target.latitude,
        target.longitude
      ),
      speed: launchedWeapon.speed,
      currentFuel: launchedWeapon.currentFuel,
      maxFuel: launchedWeapon.maxFuel,
      fuelRate: launchedWeapon.fuelRate,
      range: launchedWeapon.range,
      route: [[target.latitude, target.longitude]],
      sideColor: launchedWeapon.sideColor,
      targetId: target.id,
      lethality: launchedWeapon.lethality,
      maxQuantity: 1,
      currentQuantity: 1,
    });
    currentScenario.weapons.push(newWeapon);
  }
  launchedWeapon.currentQuantity -= launchedWeaponQuantity;
  simulationLogs.addLog(
    origin.sideId,
    `${origin.name} launched ${launchedWeaponQuantity}x ${launchedWeapon.name} at ${target.name}`,
    currentScenario.currentTime,
    SimulationLogType.WEAPON_LAUNCHED
  );
  if (launchedWeapon.currentQuantity < 1) {
    origin.weapons = origin.weapons.filter(
      (currentOriginWeapon) => currentOriginWeapon.id !== launchedWeapon.id
    );
    simulationLogs.addLog(
      origin.sideId,
      `${origin.name} has expended all stores of ${launchedWeapon.name}`,
      currentScenario.currentTime,
      SimulationLogType.WEAPON_EXPENDED
    );
  }
}

export function weaponEngagement(
  currentScenario: Scenario,
  weapon: Weapon,
  simulationLogs: SimulationLogs
) {
  if (!currentScenario.getWeapon(weapon.id)) return;

  const target =
    currentScenario.getAircraft(weapon.targetId) ??
    currentScenario.getFacility(weapon.targetId) ??
    currentScenario.getWeapon(weapon.targetId) ??
    currentScenario.getShip(weapon.targetId) ??
    currentScenario.getAirbase(weapon.targetId);
  if (target) {
    const weaponRoute = weapon.route;
    if (weaponRoute.length > 0) {
      const distanceToTargetKm = getDistanceBetweenTwoPoints(
        weapon.latitude,
        weapon.longitude,
        target.latitude,
        target.longitude
      );
      if (distanceToTargetKm <= WEAPON_ENDGAME_DISTANCE_KM) {
        weaponEndgame(currentScenario, weapon, target, simulationLogs);
        return;
      } else {
        const nextWeaponCoordinates = getNextCoordinates(
          weapon.latitude,
          weapon.longitude,
          target.latitude,
          target.longitude,
          weapon.speed
        );
        const nextWeaponLatitude = nextWeaponCoordinates[0];
        const nextWeaponLongitude = nextWeaponCoordinates[1];
        weapon.heading = getBearingBetweenTwoPoints(
          nextWeaponLatitude,
          nextWeaponLongitude,
          target.latitude,
          target.longitude
        );
        weapon.latitude = nextWeaponLatitude;
        weapon.longitude = nextWeaponLongitude;
        const remainingDistanceToTargetKm = getDistanceBetweenTwoPoints(
          weapon.latitude,
          weapon.longitude,
          target.latitude,
          target.longitude
        );
        if (remainingDistanceToTargetKm <= WEAPON_ENDGAME_DISTANCE_KM) {
          weaponEndgame(currentScenario, weapon, target, simulationLogs);
          return;
        }
      }
      weapon.currentFuel -= weapon.fuelRate / 3600;
      if (weapon.currentFuel <= 0) {
        currentScenario.weapons = currentScenario.weapons.filter(
          (currentScenarioWeapon) => currentScenarioWeapon.id !== weapon.id
        );
        simulationLogs.addLog(
          weapon.sideId,
          `${weapon.name} has run out of fuel and is no longer operational`,
          currentScenario.currentTime,
          SimulationLogType.WEAPON_CRASHED
        );
      }
    }
  } else {
    currentScenario.weapons = currentScenario.weapons.filter(
      (currentScenarioWeapon) => currentScenarioWeapon.id !== weapon.id
    );
    simulationLogs.addLog(
      weapon.sideId,
      `${weapon.name} has lost its target and is no longer operational`,
      currentScenario.currentTime,
      SimulationLogType.WEAPON_CRASHED
    );
  }
}

export function aircraftPursuit(currentScenario: Scenario, aircraft: Aircraft) {
  const target = currentScenario.getAircraft(aircraft.targetId);
  if (!target) {
    aircraft.targetId = "";
    return;
  }
  if (aircraft.weapons.length < 1) return;

  const TRAIL_DISTANCE_NM = 5;
  const trailKm = (TRAIL_DISTANCE_NM * NAUTICAL_MILES_TO_METERS) / 1000;
  const behindBearing = (target.heading + 180) % 360;
  const trailPosition = getTerminalCoordinatesFromDistanceAndBearing(
    target.latitude,
    target.longitude,
    trailKm,
    behindBearing
  );
  const trailLat = trailPosition[0];
  const trailLon = trailPosition[1];

  aircraft.route = [[trailLat, trailLon]];
  aircraft.heading = getBearingBetweenTwoPoints(
    aircraft.latitude,
    aircraft.longitude,
    trailLat,
    trailLon
  );
}

export function routeAircraftToStrikePosition(
  currentScenario: Scenario,
  aircraft: Aircraft,
  targetId: string,
  strikeRadiusNm: number
) {
  const target =
    currentScenario.getFacility(targetId) ||
    currentScenario.getShip(targetId) ||
    currentScenario.getAirbase(targetId) ||
    currentScenario.getAircraft(targetId);
  if (!target) return;
  if (aircraft.weapons.length < 1) return;

  const bearingBetweenAircraftAndTarget = getBearingBetweenTwoPoints(
    aircraft.latitude,
    aircraft.longitude,
    target.latitude,
    target.longitude
  );
  const bearingBetweenTargetAndAircraft = getBearingBetweenTwoPoints(
    target.latitude,
    target.longitude,
    aircraft.latitude,
    aircraft.longitude
  );
  const strikeLocation = getTerminalCoordinatesFromDistanceAndBearing(
    target.latitude,
    target.longitude,
    (strikeRadiusNm * NAUTICAL_MILES_TO_METERS) / 1000,
    bearingBetweenTargetAndAircraft
  );

  aircraft.route.push([strikeLocation[0], strikeLocation[1]]);
  aircraft.heading = bearingBetweenAircraftAndTarget;
}

File: gym/blade/engine/weaponEngagement.py
from blade.units.Aircraft import Aircraft
from blade.units.Ship import Ship
from blade.units.Facility import Facility
from blade.units.Airbase import Airbase
from blade.units.Weapon import Weapon
from blade.Scenario import Scenario
from uuid import uuid4

from blade.utils.constants import NAUTICAL_MILES_TO_METERS
from blade.utils.utils import (
    get_bearing_between_two_points,
    get_distance_between_two_points,
    get_next_coordinates,
    get_terminal_coordinates_from_distance_and_bearing,
    random_float,
    random_int,
)
from blade.engine.electronicWarfare import get_effective_detection_range
from blade.engine.environmentConstraints import get_environment_detection_range

Target = Aircraft | Facility | Weapon | Airbase | Ship

UNIT_SCORE_TABLE = {
    "aircraft": 10,
    "weapon": 1,
    "facility": 30,
    "ship": 50,
    "airbase": 100,
}
OBJECTIVE_BONUS_SCORE = 200


def _target_type(target: Target) -> str:
    if isinstance(target, Aircraft):
        return "aircraft"
    if isinstance(target, Facility):
        return "facility"
    if isinstance(target, Airbase):
        return "airbase"
    if isinstance(target, Ship):
        return "ship"
    return "weapon"


def _expand_target_type(target_type: str) -> set[str]:
    normalized = target_type.strip().lower().replace("-", "_")
    if normalized in {"*", "any", "all"}:
        return {"aircraft", "facility", "airbase", "ship", "weapon"}
    if normalized in {"air", "aircraft"}:
        return {"aircraft"}
    if normalized in {"missile", "weapon", "munition"}:
        return {"weapon"}
    if normalized in {"surface", "ground", "land"}:
        return {"facility", "airbase"}
    if normalized in {"sea", "maritime", "naval", "ship"}:
        return {"ship"}
    if normalized in {"fixed", "fixed_site", "base"}:
        return {"facility", "airbase"}
    return {normalized}


def _infer_weapon_target_types(weapon: Weapon) -> set[str]:
    explicit_types = getattr(weapon, "target_types", None) or []
    if explicit_types:
        expanded: set[str] = set()
        for target_type in explicit_types:
            expanded.update(_expand_target_type(str(target_type)))
        return expanded

    class_name = str(getattr(weapon, "class_name", "") or "").lower()

    if class_name.startswith("aim-") or "sidewinder" in class_name or "amraam" in class_name:
        return {"aircraft"}
    if "harpoon" in class_name:
        return {"ship"}
    if class_name.startswith("agm-") or "tomahawk" in class_name or "jassm" in class_name or "alcm" in class_name:
        return {"facility", "airbase", "ship"}
    if (
        class_name.startswith("rim-")
        or class_name.startswith("hq-")
        or class_name.startswith("48n6")
        or class_name.startswith("9m")
        or class_name.startswith("57e6")
        or "s-300" in class_name
        or "s-400" in class_name
        or "s-500" in class_name
        or "buk" in class_name
        or "tor-" in class_name
        or "pantsir" in class_name
        or "patriot" in class_name
        or "thaad" in class_name
        or "aster" in class_name
        or "barak" in class_name
        or "nasams" in class_name
        or "standard" in class_name
        or "ram" in class_name
    ):
        return {"aircraft", "weapon"}

    return {"aircraft", "facility", "airbase", "ship", "weapon"}


def weapon_can_target_type(target: Target, weapon: Weapon) -> bool:
    return _target_type(target) in _infer_weapon_target_types(weapon)


def is_threat_detected(
    threat: Target,
    detector: Facility | Ship | Aircraft,
    current_scenario: Scenario | None = None,
) -> bool:
    distance_to_threat_km = get_distance_between_two_points(
        detector.latitude,
        detector.longitude,
        threat.latitude,
        threat.longitude,
    )
    distance_to_threat_nm = (
        distance_to_threat_km * 1000
    ) / NAUTICAL_MILES_TO_METERS
    detection_range_nm = (
        get_effective_detection_range(current_scenario, detector)
        if current_scenario is not None
        else detector.get_detection_range()
    )
    if current_scenario is not None:
        detection_range_nm = get_environment_detection_range(
            current_scenario, detector, threat, detection_range_nm
        )
    return distance_to_threat_nm <= detection_range_nm


def _remove_if_present(items: list, item) -> None:
    if item in items:
        items.remove(item)


def weapon_can_engage_target(target: Target, weapon: Weapon) -> bool:
    if not weapon_can_target_type(target, weapon):
        return False

    weapon_engagement_range_nm = weapon.get_engagement_range()

    distance_to_target_km = get_distance_between_two_points(
        weapon.latitude, weapon.longitude, target.latitude, target.longitude
    )

    distance_to_target_nm = (distance_to_target_km * 1000) / NAUTICAL_MILES_TO_METERS

    return distance_to_target_nm <= weapon_engagement_range_nm


def check_target_tracked_by_count(current_scenario: Scenario, target: Target) -> int:
    count = 0
    for weapon in current_scenario.weapons:
        if weapon.target_id == target.id:
            count += 1
    return count


def on_unit_destroyed(
    current_scenario: Scenario, attacker_side_id: str, target: Target
) -> None:
    attacker_side = current_scenario.get_side(attacker_side_id)
    if attacker_side is None:
        return

    base_score = UNIT_SCORE_TABLE["weapon"]
    unit_type = "weapon"
    if isinstance(target, Aircraft):
        base_score = UNIT_SCORE_TABLE["aircraft"]
        unit_type = "aircraft"
    elif isinstance(target, Facility):
        base_score = UNIT_SCORE_TABLE["facility"]
        unit_type = "facility"
    elif isinstance(target, Ship):
        base_score = UNIT_SCORE_TABLE["ship"]
        unit_type = "ship"
    elif isinstance(target, Airbase):
        base_score = UNIT_SCORE_TABLE["airbase"]
        unit_type = "airbase"

    is_objective = not isinstance(target, Weapon) and getattr(
        target, "is_objective", False
    )
    attacker_side.total_score = getattr(attacker_side, "total_score", 0) + base_score
    if is_objective:
        attacker_side.total_score += OBJECTIVE_BONUS_SCORE
        current_scenario.last_objective_destroyed = {
            "attacker_side_id": attacker_side_id,
            "victim_side_id": getattr(target, "side_id", ""),
            "unit_id": target.id,
            "unit_name": target.name,
            "unit_type": unit_type,
            "destroyed_at": current_scenario.current_time,
        }


def weapon_endgame(current_scenario: Scenario, weapon: Weapon, target: Target) -> bool:
    _remove_if_present(current_scenario.weapons, weapon)
    if random_float(0, 1) <= weapon.lethality:
        if isinstance(target, Aircraft):
            _remove_if_present(current_scenario.aircraft, target)
        elif isinstance(target, Ship):
            _remove_if_present(current_scenario.ships, target)
        elif isinstance(target, Facility):
            _remove_if_present(current_scenario.facilities, target)
        elif isinstance(target, Airbase):
            _remove_if_present(current_scenario.airbases, target)
        elif isinstance(target, Weapon):
            _remove_if_present(current_scenario.weapons, target)
        on_unit_destroyed(current_scenario, weapon.side_id, target)
        return True
    return False


def launch_weapon(
    current_scenario: Scenario,
    origin: Facility | Ship | Aircraft,
    target: Target,
    launched_weapon: Weapon,
    launched_weapon_quantity: int,
) -> None:
    if (
        launched_weapon_quantity <= 0
        or len(origin.weapons) == 0
        or launched_weapon.current_quantity < launched_weapon_quantity
        or launched_weapon.speed <= 0
    ):
        return

    for _ in range(launched_weapon_quantity):
        next_weapon_coordinates = get_next_coordinates(
            origin.latitude,
            origin.longitude,
            target.latitude,
            target.longitude,
            launched_weapon.speed,
        )
        next_weapon_latitude = next_weapon_coordinates[0]
        next_weapon_longitude = next_weapon_coordinates[1]
        new_weapon = Weapon(
            id=str(uuid4()),
            name=f"{launched_weapon.name} #{random_int(0, 1000)}",
            side_id=origin.side_id,
            class_name=launched_weapon.class_name,
            latitude=next_weapon_latitude,
            longitude=next_weapon_longitude,
            altitude=launched_weapon.altitude,
            heading=get_bearing_between_two_points(
                next_weapon_latitude,
                next_weapon_longitude,
                target.latitude,
                target.longitude,
            ),
            speed=launched_weapon.speed,
            current_fuel=launched_weapon.current_fuel,
            max_fuel=launched_weapon.max_fuel,
            fuel_rate=launched_weapon.fuel_rate,
            range=launched_weapon.range,
            route=[[target.latitude, target.longitude]],
            side_color=launched_weapon.side_color,
            target_id=target.id,
            lethality=launched_weapon.lethality,
            current_quantity=1,
            max_quantity=1,
            target_types=getattr(launched_weapon, "target_types", []),
        )
        current_scenario.weapons.append(new_weapon)
    launched_weapon.current_quantity -= launched_weapon_quantity
    if launched_weapon.current_quantity < 1:
        origin.weapons.remove(launched_weapon)


def weapon_engagement(current_scenario: Scenario, weapon: Weapon) -> None:
    target = current_scenario.get_target(weapon.target_id)
    if target is None:
        _remove_if_present(current_scenario.weapons, weapon)
    else:
        weapon_route = weapon.route
        if len(weapon_route) > 0:
            # there is a weird bug where a weapon will be teleported a vast distance if it gets too close to the target but weaponEndgame is not called, current solution is to set threshold to 1 km
            if (
                get_distance_between_two_points(
                    weapon.latitude,
                    weapon.longitude,
                    target.latitude,
                    target.longitude,
                )
                < 1
            ):
                weapon_endgame(current_scenario, weapon, target)
            else:
                next_weapon_coordinates = get_next_coordinates(
                    weapon.latitude,
                    weapon.longitude,
                    target.latitude,
                    target.longitude,
                    weapon.speed,
                )
                next_weapon_latitude = next_weapon_coordinates[0]
                next_weapon_longitude = next_weapon_coordinates[1]
                weapon.heading = get_bearing_between_two_points(
                    next_weapon_latitude,
                    next_weapon_longitude,
                    target.latitude,
                    target.longitude,
                )
                weapon.latitude = next_weapon_latitude
                weapon.longitude = next_weapon_longitude
                weapon.current_fuel -= weapon.fuel_rate / 3600
                if weapon.current_fuel <= 0:
                    _remove_if_present(current_scenario.weapons, weapon)


def aircraft_pursuit(
    current_scenario: Scenario,
    aircraft: Aircraft,
) -> None:
    target = current_scenario.get_aircraft(aircraft.target_id)
    if target is None:
        aircraft.target_id = ""
        return
    if len(aircraft.weapons) < 1:
        return
    
    TRAIL_DISTANCE_NM = 5
    trail_km = (TRAIL_DISTANCE_NM * NAUTICAL_MILES_TO_METERS) / 1000
    behind_bearing = (target.heading + 180) % 360
    trail_position = get_terminal_coordinates_from_distance_and_bearing(
        target.latitude,
        target.longitude,
        trail_km,
        behind_bearing,
    )
    trail_latitude = trail_position[0]
    trail_longitude = trail_position[1]

    aircraft.route = [
        [trail_latitude, trail_longitude],
    ]
    aircraft.heading = get_bearing_between_two_points(
        aircraft.latitude,
        aircraft.longitude,
        trail_latitude,
        trail_longitude,
    )


def route_aircraft_to_strike_position(
    current_scenario: Scenario,
    aircraft: Aircraft,
    target_id: str,
    strike_radius_nm: float,
) -> None:
    target = current_scenario.get_target(target_id)
    if target is None:
        return
    if len(aircraft.weapons) < 1:
        return

    bearing_between_aircraft_and_target = get_bearing_between_two_points(
        aircraft.latitude, aircraft.longitude, target.latitude, target.longitude
    )
    bearing_between_target_and_aircraft = get_bearing_between_two_points(
        target.latitude, target.longitude, aircraft.latitude, aircraft.longitude
    )
    strike_location = get_terminal_coordinates_from_distance_and_bearing(
        target.latitude,
        target.longitude,
        (strike_radius_nm * NAUTICAL_MILES_TO_METERS) / 1000,
        bearing_between_target_and_aircraft,
    )

    aircraft.route.append([strike_location[0], strike_location[1]])
    aircraft.heading = bearing_between_aircraft_and_target
